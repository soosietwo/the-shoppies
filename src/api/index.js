const getMovies = async (searchTerm, year, page = 1) => {
  let url = `http://www.omdbapi.com/?apikey=${process.env.REACT_APP_API_KEY}&type=movie&s=${searchTerm}&page=${page}`;

  if (year) {
    url = url.concat(`&y=${year}`);
  }

  return (await fetch(url)).json();
};

export default {
  getMovies,
};
