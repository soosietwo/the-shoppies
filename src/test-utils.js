// test-utils.js
import React from "react";
import { render } from "@testing-library/react";
import { AppProvider, Frame } from "@shopify/polaris";
import en from "@shopify/polaris/locales/en.json";
import Store from "./store";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const AllTheProviders = ({ children }) => {
  return (
    <AppProvider i18n={en}>
      <Store>
        <Frame>{children}</Frame>
      </Store>
    </AppProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
