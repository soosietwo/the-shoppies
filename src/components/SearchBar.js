import React, { useCallback, useState, useContext } from "react";
import {
  Form,
  FormLayout,
  TextField,
  Button,
  Icon,
  Stack,
} from "@shopify/polaris";
import { SearchMajorMonotone } from "@shopify/polaris-icons";

import api from "../api";
import { Context } from "../store";
import {
  FETCH_MOVIES,
  FETCH_MOVIES_SUCCESS,
  FETCH_MOVIES_FAILURE,
} from "../store/constants";

function SearchBar(props) {
  const [state, dispatch] = useContext(Context);
  const [searchTerm, setSearchTerm] = useState("");
  const [year, setYear] = useState("");

  const handleSearchChange = useCallback((value) => setSearchTerm(value), []);
  const handleYearChange = useCallback((value) => setYear(value), []);

  const handleSubmit = () => {
    dispatch({ type: FETCH_MOVIES, searchTerm });

    api
      .getMovies(searchTerm, year, 1)
      .then(({ Error, Search, totalResults }) => {
        if (Error) {
          dispatch({ type: FETCH_MOVIES_FAILURE, payload: Error });
        } else {
          dispatch({
            type: FETCH_MOVIES_SUCCESS,
            payload: { Search, totalResults },
          });
        }
      })
      .catch((error) =>
        dispatch({ type: FETCH_MOVIES_FAILURE, payload: error })
      );
  };

  return (
    <Form onSubmit={handleSubmit}>
      <FormLayout>
        <Stack>
          <Stack.Item fill>
            <TextField
              label="Search for a movie title"
              labelHidden
              fill
              type="text"
              value={searchTerm}
              prefix={<Icon source={SearchMajorMonotone} />}
              onChange={handleSearchChange}
            />
          </Stack.Item>
          {/* <TextField
          label="Year"
          type="text"
          value={year}
          onChange={handleYearChange}
        /> */}
          <Button
            primary
            disabled={!searchTerm.trim().length}
            loading={state.isLoading}
            submit
          >
            Search movies
          </Button>
        </Stack>
      </FormLayout>
    </Form>
  );
}

export default SearchBar;
