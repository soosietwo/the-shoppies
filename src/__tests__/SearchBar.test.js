import React from "react";
import { render, waitFor } from "../test-utils";
import SearchBar from "../components/SearchBar";
import { fireEvent } from "@testing-library/react/dist/pure";
import api from "../api";
import Store from "../store";

describe("SearchBar", () => {
  it("renders and matches snapshot", () => {
    const { container } = render(
      <Store>
        <SearchBar />
      </Store>
    );

    expect(container).toMatchSnapshot();
  });

  it("calls getMovies with search term on submit", async () => {
    const mockInitialState = {
      currentPage: 1,
      totalResults: 0,
    };

    const { container } = render(
      <Store initialState={mockInitialState}>
        <SearchBar />
      </Store>
    );

    const getMoviesSpy = jest.spyOn(api, "getMovies");

    fireEvent.change(container.querySelector("#PolarisSearchField1"), {
      target: { value: "test" },
    });

    await waitFor(() => {
      expect(getMoviesSpy).toHaveBeenCalledWith("test", 1);
    });
  });
});
