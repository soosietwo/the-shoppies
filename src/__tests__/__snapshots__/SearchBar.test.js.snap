// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders and matches snapshot 1`] = `
<div>
  <div>
    <div
      class="Polaris-Frame"
      data-polaris-layer="true"
    >
      <div
        class="Polaris-Frame__Skip"
      >
        <a
          class="Polaris-Button"
          data-polaris-unstyled="true"
          data-polaris-unstyled-button="true"
          href="#AppFrameMainContent"
        >
          <span
            class="Polaris-Button__Content"
          >
            <span
              class="Polaris-Button__Text"
            >
              Skip to content
            </span>
          </span>
        </a>
      </div>
      <div
        class="Polaris-Frame__ContextualSaveBar Polaris-Frame-CSSAnimation--startFade"
      />
      <main
        class="Polaris-Frame__Main"
        data-has-global-ribbon="false"
        id="AppFrameMain"
      >
        <a
          id="AppFrameMainContent"
          tabindex="-1"
        />
        <div
          class="Polaris-Frame__Content"
        >
          <div
            class="Polaris-TopBar-SearchField"
          >
            <span
              class="Polaris-VisuallyHidden"
            >
              <label
                for="PolarisSearchField1"
              >
                Search
              </label>
            </span>
            <input
              autocapitalize="off"
              autocomplete="off"
              autocorrect="off"
              class="Polaris-TopBar-SearchField__Input"
              id="PolarisSearchField1"
              placeholder="Search"
              type="search"
              value=""
            />
            <span
              class="Polaris-TopBar-SearchField__Icon"
            >
              <span
                class="Polaris-Icon"
              >
                <svg
                  aria-hidden="true"
                  class="Polaris-Icon__Svg"
                  focusable="false"
                  viewBox="0 0 20 20"
                >
                  <path
                    d="M8 12a4 4 0 1 1 0-8 4 4 0 0 1 0 8zm9.707 4.293l-4.82-4.82A5.968 5.968 0 0 0 14 8 6 6 0 0 0 2 8a6 6 0 0 0 6 6 5.968 5.968 0 0 0 3.473-1.113l4.82 4.82a.997.997 0 0 0 1.414 0 .999.999 0 0 0 0-1.414z"
                  />
                </svg>
              </span>
            </span>
            <div
              class="Polaris-TopBar-SearchField__Backdrop Polaris-TopBar-SearchField__BackdropShowFocusBorder"
            />
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
